lib/date_range.rb

```diff
@@ -0,0 +1,34 @@
+require 'date'
+
+module Hotel
+  class DateRange
+    attr_accessor :start_date, :end_date
+
+    def initialize(start_date, end_date)
+      raise ArgumentError.new("start date has to be at least 1 day before end date") if (end_date - start_date).to_i < 1
+      @start_date = start_date
+      @end_date = end_date
+    end
+
+    def overlap?(range)
+      self.nights.times do |day|
+        range.nights.times do |day_in_range|
+          return true if (range.start_date + day_in_range) == (@start_date + day)
+        end
+      end
+      return false
```

Comment:
There is a simpler version of this logic that doesn't require looping! Draw a timeline, and put a date range on it, then draw out all possible relationships the ranges can have with it. I often take the approach of making a table or drawing a sketch when I want to simplify boolean logic.